libraries/Schedule/src/circular_queue/circular_queue.h

```diff
@@ -0,0 +1,281 @@
+        @return The number of elements actually copied into the queue, counted
+                from the buffer head.
+    */
+    size_t push_n(const T* buffer, size_t size)
+    {
+        const auto inPos = m_inPos.load(std::memory_order_acquire);
+        const auto outPos = m_outPos.load(std::memory_order_relaxed);
+
+        size_t blockSize = (outPos > inPos) ? outPos - 1 - inPos : (outPos == 0) ? m_bufSize - 1 - inPos : m_bufSize - inPos;
+        blockSize = std::min(size, blockSize);
+        if (!blockSize) return 0;
+        int next = (inPos + blockSize) % m_bufSize;
+
+        std::atomic_thread_fence(std::memory_order_acquire);
+
+        auto dest = m_buffer.get() + inPos;
+        std::copy_n(std::make_move_iterator(buffer), blockSize, dest);
+        size = std::min(size - blockSize, outPos > 1 ? static_cast<size_t>(outPos - next - 1) : 0);
+        next += size;
+        dest = m_buffer.get();
+        std::copy_n(std::make_move_iterator(buffer + blockSize), size, dest);
+
+        std::atomic_thread_fence(std::memory_order_release);
+
+        m_inPos.store(next, std::memory_order_release);
+        return blockSize + size;
+    }
+
+    /*!
+        @brief	Pop the next available element from the queue.
+        @return An rvalue copy of the popped element, or a default
+                value of type T if the queue is empty.
+    */
+    T pop()
+    {
+        const auto outPos = m_outPos.load(std::memory_order_acquire);
+        if (m_inPos.load(std::memory_order_relaxed) == outPos) return defaultValue;
+
+        std::atomic_thread_fence(std::memory_order_acquire);
+
+        auto val = std::move(m_buffer[outPos]);
+
+        std::atomic_thread_fence(std::memory_order_release);
+
+        m_outPos.store((outPos + 1) % m_bufSize, std::memory_order_release);
+        return val;
+    }
+
+    /*!
+        @brief	Pop multiple elements in ordered sequence from the queue to a buffer.
+        @return The number of elements actually popped from the queue to
+                buffer.
+    */
+    size_t pop_n(T* buffer, size_t size) {
+        size_t avail = size = std::min(size, available());
+        if (!avail) return 0;
+        const auto outPos = m_outPos.load(std::memory_order_acquire);
+        size_t n = std::min(avail, static_cast<size_t>(m_bufSize - outPos));
+
+        std::atomic_thread_fence(std::memory_order_acquire);
+
+        buffer = std::copy_n(std::make_move_iterator(m_buffer.get() + outPos), n, buffer);
+        avail -= n;
+        std::copy_n(std::make_move_iterator(m_buffer.get()), avail, buffer);
+
+        std::atomic_thread_fence(std::memory_order_release);
+
+        m_outPos.store((outPos + size) % m_bufSize, std::memory_order_release);
+        return size;
+    }
+
+    /*!
+        @brief	Iterate over and remove each available element from queue,
+                calling back fun with an rvalue reference of every single element.
+    */
+    void for_each(std::function<void(T&&)> fun)
```

Comment:
Instead of polluting the class with an iteration method, how about defining an iterator class and making it compatible with STL algorithms? That would decouple the iteration logic from the class itself, which is at it should be.

Reply:
E.g.: add begin(), end(), iterator class, const_iterator class, etc. Then the user could do e.g.: `for(auto t : myq) {fun(t);}` or whatever. I would really prefer such compatibility, I was even thinking why not implement a circular queue in terms of an existing STL container, but I'm ok with a custom implementation if it's compatible.

Reply:
Uh oh - I am not sure that this can really be the same logic as the `for_each()`, more specifically the `for_each_requeue()`. There might a way to do this in STL style, but that looks like a major piece of overhaul, and research. As I am actively thinking about doing more special stuff with interaction and lambda IOC, I'd rather no be bound by compatibility constraints (just now). There are some performance problems that the linked list doesn't suffer, I would rather try and invent something to get closer in performance to that, next.